### ISSUE TITLE
Possible rate limit or other issue

### ISSUE THREAD

**Issue by OGProgrammer** [adamdecaf/terraform-provider-namecheap#9]
If you run `terraform plan` back-to-back, it seems Namecheap will throttle your requests. I get the following error (in summary) from terraform:

`expected element type <ApiResponse> but have <html>`

### Terraform Version
Terraform v0.11.10
Namecheap v1.1.1

### Affected Resource(s)
- namecheap_record
- namecheap_ns

### Debug Output

Terraform trace output towards the end of a `terraform apply` I executed back to back a few times:

```
2018/12/09 12:30:04 [TRACE] [walkRefresh] Exiting eval tree: namecheap_record.copelander-com
2018/12/09 12:30:05 [ERROR] root: eval: *terraform.EvalRefresh, err: namecheap_record.www-copelander-com: Couldn't find namecheap record: expected element type <ApiResponse> but have <html>
2018/12/09 12:30:05 [ERROR] root: eval: *terraform.EvalSequence, err: namecheap_record.www-copelander-com: Couldn't find namecheap record: expected element type <ApiResponse> but have <html>
2018/12/09 12:30:05 [TRACE] [walkRefresh] Exiting eval tree: namecheap_record.www-copelander-com
2018/12/09 12:30:05 [ERROR] root: eval: *terraform.EvalRefresh, err: namecheap_record.www-remotedevforce-com: Couldn't find namecheap record: expected element type <ApiResponse> but have <html>
2018/12/09 12:30:05 [ERROR] root: eval: *terraform.EvalSequence, err: namecheap_record.www-remotedevforce-com: Couldn't find namecheap record: expected element type <ApiResponse> but have <html>
2018/12/09 12:30:05 [TRACE] [walkRefresh] Exiting eval tree: namecheap_record.www-remotedevforce-com
2018/12/09 12:30:05 [TRACE] dag/walk: upstream errored, not walking "provider.namecheap (close)"
2018/12/09 12:30:05 [TRACE] dag/walk: upstream errored, not walking "root"

2018/12/09 12:30:05 [DEBUG] plugin: waiting for all plugin processes to complete...
Error: Error refreshing state: 3 error(s) occurred:

* namecheap_record.copelander-com: 1 error(s) occurred:

2018-12-09T12:30:05.199-0800 [DEBUG] plugin.terraform-provider-namecheap: 2018/12/09 12:30:05 [ERR] plugin: plugin server: accept unix /var/folders/03/blcr90b52jb9d_4qphynlz680000gn/T/plugin407438253: use of closed network connection
2018-12-09T12:30:05.199-0800 [DEBUG] plugin.terraform-provider-google_v1.19.1_x4: 2018/12/09 12:30:05 [ERR] plugin: plugin server: accept unix /var/folders/03/blcr90b52jb9d_4qphynlz680000gn/T/plugin922996050: use of closed network connection
* namecheap_record.copelander-com: namecheap_record.copelander-com: Couldn't find namecheap record: expected element type <ApiResponse> but have <html>
* namecheap_record.www-copelander-com: 1 error(s) occurred:

* namecheap_record.www-copelander-com: namecheap_record.www-copelander-com: Couldn't find namecheap record: expected element type <ApiResponse> but have <html>
* namecheap_record.www-remotedevforce-com: 1 error(s) occurred:

* namecheap_record.www-remotedevforce-com: namecheap_record.www-remotedevforce-com: Couldn't find namecheap record: expected element type <ApiResponse> but have <html>


2018-12-09T12:30:05.202-0800 [DEBUG] plugin: plugin process exited: path=/Users/me/my_project/.terraform/plugins/darwin_amd64/terraform-provider-google_v1.19.1_x4
2018-12-09T12:30:05.203-0800 [DEBUG] plugin: plugin process exited: path=/Users/me/.terraform.d/plugins/terraform-provider-namecheap
Error: terraform apply failed.
```

### Expected Behavior
I know when [AWS hits rate limits](https://github.com/aws/aws-sdk-go/blob/09642736477d485f9ec9ad5754a945123923eb96/aws/request/retryer.go#L39), terraform will exponentially back-off from hitting the API too much. I'd expect if we hit any rate limits with namecheap, we would back-off, and retry. Unfortunately they are not returning anything in their headers I can see for rate limiting which I know some APIs do. 

### Actual Behavior
Namecheap API seems to bomb out somehow. Getting back this error `expected element type <ApiResponse> but have <html>`

### Steps to Reproduce
If you have a dozen of so records managed by terraform with namecheap. Run:
1. `terraform plan` over and over till the API bombs out.

### References

Maybe it has something to do with rate limiting? Still don't know why it says "html" though. A capped out response from namecheap looks like this for me hitting getList dns over and over with my browser.
```
<ApiResponse Status="ERROR">
<Errors>
<Error Number="500000">Too many requests</Error>
</Errors>
<Server/>
<GMTTimeDifference/>
<ExecutionTime>0</ExecutionTime>
</ApiResponse>
```

Maybe there is a way to account for this (if this is the issue I guess, I'm still assuming this has to do with rate limiting since the first few `terraform plan` jobs will work fine until it bombs out) and "backoff of the API" then retry like AWS and other providers do?

Thanks again for these great go libraries.

**Comment by adamdecaf** [comment]
Makes sense. I pushed off a backoff aware client until it was needed. Do you know if they return a header regarding the rate limit? 

We can look to see if there are better calls to minimize the requests too.

**Comment by adamdecaf** [comment]
Regarding a retry/backoff aware `http.Client` I'd recommend using https://github.com/hashicorp/go-retryablehttp - thoughts? We should be able to hook in a `Backoff` func to scrape the XML or a header value.

**Comment by OGProgrammer** [comment]
This is very much an RPC style API... response is a 200 even on error. No rate limiting headers in the response before or during a throttle. Not sure if we should ask support about what the limits are because I don't see anything in the docs. That library looks good, maybe if we see ApiResponse  Status="ERROR" with the Error # 500000 then we start backing off a bit. I'm going to take a look if there are examples out there of retryablehttp being used in a tf provider.

The response we can parse:
```
<ApiResponse Status="ERROR">
<Errors>
<Error Number="500000">Too many requests</Error>
</Errors>
```

**Comment by OGProgrammer** [comment]
If we also figure out what the response is when the ip isn't whitelisted, perhaps we can provide better feedback on that. Right now if you are on a new ip, you get back `namecheap_record.example-com: namecheap_record.example-com: Couldn't find namecheap record: Could not find the record` which could be confusing for a dev seeing this and not realizing it is whitelist ip related.

**Comment by OGProgrammer** [comment]
Yo @adamdecaf , this pissed me off enough to fix it. Checkout the PR https://github.com/adamdecaf/terraform-provider-namecheap/pull/11

Looks like it is working well thus far. Appreciate your feedback, wondering a bit about the mutexes and if I messed anything up there.

**Comment by adamdecaf** [comment]
@OGProgrammer Can you try out `master`? It worked for me to create a record, so I hope the retries work for you also! If so I'll make a proper release.

**Comment by OGProgrammer** [comment]
Tested `master` and it so slowww. I think you need to double check whats up with your branch. The branch I have `hotfix/rateThrottling` is blazing fast, and `master` seems to cause issues where namecheap returns some error. If you could pass the error up somehow through the `namecheap` go depenedency then we could get more visibility maybe?

Not sure what the issue is but I'm sticking with my branch in the meantime.

**Comment by adamdecaf** [comment]
@OGProgrammer Whoops. I've fixed that in https://github.com/adamdecaf/terraform-provider-namecheap/commit/bab5a379587159efacae63b2bf52653664c82707. It works for me to create a record in 6-9s and delete one in ~6s. Hope it's fast for you too now!

**Comment by OGProgrammer** [comment]
Looks good now!

**Comment by adamdecaf** [comment]
Awesome! I released 1.2.0 which has binaries to download 

https://github.com/adamdecaf/terraform-provider-namecheap/releases/tag/1.2.0

Let me know if you have other issues, but thanks for the help on this!
